### PR TITLE
Convert GMP BigInt class to record

### PR DESCRIPTION
**Pull request by Ldelaney**
- Changes the BigInt class (which wraps the GMP mpz multiple-precision-integers) into a record. 
- Adds operator overloads for BigInt
- Also adds testing for the GMP module.
- As a result of this testing, several bugs with the GMP module were fixed, including ones where a function extern had a typo or incorrect number/type of arguments. A few functions were renamed or moved out of the record scope to be consistent with the naming and placement of other functions. (functions that return an integer and do not modify any of their operands are defined in the module outside of the record scope).

### CODE REVIEW

File: modules/standard/GMP.chpl [hunk 1]
```diff
@@ -93,12 +90,12 @@ A code example::
 
  writeln(a); // output 100!
 
- delete a; // free memory used by the GMP value
+ // memory used by the GMP value is freed automatically by record destruction
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: I know this is in a comment but there's a bad line break here.

File: modules/standard/GMP.chpl [hunk 2]
```diff
@@ -152,7 +148,7 @@ module GMP {
   extern type __mpz_struct;
 
   /* The GMP ``mpz_t`` type */
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: This trailing comment does not provide any information.  Remove or expand.

File: modules/standard/GMP.chpl [hunk 3]
```diff
@@ -462,110 +451,142 @@ module GMP {
...
+      if !(this.owned) { //TODO: confirm that this case gets tested
+        on Locales[this.locale_id] do mpz_init(this.mpz);
+        this.owned = true;
+      }
+
+      // if we don't need to deep copy, free ours, make a shallow copy, get out
+      if !needToCopy {
+        on Locales[this.locale_id] do mpz_clear(this.mpz);
+        this.owned = false;
+        this.mpz = num;
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: This return is unnecessary

File: modules/standard/GMP.chpl [hunk 4]
```diff
@@ -462,110 +451,142 @@ module GMP {
...
-        var mpz_struct = this.mpz[1];
-        var tmp = new BigInt(true, (mp_bits_per_limb:uint(64))*chpl_gmp_mpz_nlimbs(mpz_struct));
-        chpl_gmp_get_mpz(tmp.mpz, this.locale.id, mpz_struct); 
-        return (true, tmp);
-      }
-    }
-
-    // Assignment functions
-    proc set(a:BigInt)
+   // Assignment functions
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: Probably a topic for face-to-face

I am not sure what is happening to the existing state of the current BigInt for owned == true vs owned == false

File: modules/standard/GMP.chpl [hunk 5]
```diff
@@ -462,110 +451,142 @@ module GMP {
...
+    proc set(ref a: BigInt)
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: Can formal <a> be declared with intent "const ref" rather than ref?

File: modules/standard/GMP.chpl [hunk 6]
```diff
@@ -581,838 +602,993 @@ module GMP {
...
+        if (here.id != a.locale_id) {
+          var a_ = a;
+          mpz_sub_ui(this.mpz, a_.mpz, b.safeCast(c_ulong));
+        } else {
+          mpz_sub_ui(this.mpz, a.mpz, b.safeCast(c_ulong));
+        }
       }
     }
-    proc ui_sub(a:uint, b:BigInt)
+    proc ui_sub(a: uint, ref b: BigInt)
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: For face-to-face

I don't have any reason to believe there's a problem here but I'd like to work through the
mechanisms that enable the case for here.id != b.locale_id to function correctly i.e. the
deep copy required, any tests, etc.

File: modules/standard/GMP.chpl [hunk 7]
```diff
@@ -1421,179 +1597,565 @@ module GMP {
...
-    //gmp_asprintf(s, "%Zd", a_.mpz);
+    var s: string;
+    on Locales[this.locale_id] {
+      s = this.get_str();
+    }
     writer.write(s);
-    if acopy then delete a_;
   }
 
+  // Operator Overloads
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: Why are all of these declared inline?  This is going to be slow operation.  Shame on us if
the call overhead matters.

File: modules/standard/GMP.chpl [hunk 8]
```diff
@@ -581,838 +602,993 @@ module GMP {
...
+      on Locales[this.locale_id] {
+        var tmpvar = chpl_gmp_mpz_get_str(base.safeCast(c_int), this.mpz);
+        ret = new string(tmpvar, owned=true, needToCopy=false);
       }
       return ret;
     }
 
     // Arithmetic functions
-    proc add(a:BigInt, b:BigInt)
+    proc add(ref a: BigInt, ref b: BigInt)
```
Comment (noakesmichael) [inline modules/standard/GMP.chpl]: If I am reading this correctly this implements

  this = a + b;

If this is right then

1) I think that both formals should be const ref (which is the supposed to be the default intent but might in fact behave differently if the blank intent is actually used)

2) I am unsure how to think about this method.  I see that it is used to support

  inline proc +(ref a: BigInt, ref b: BigInt){
    var c = new BigInt();
    c.add(a, b);
    return c;
  }

but wonder what other roles it is supposed to play or if this is the right way to think
about the desired operator.  I will cogitate on this for our face-to-face

File: modules/standard/GMP_old.chpl [hunk 9]
```diff
@@ -0,0 +1,1723 @@
+/*
```
Comment (noakesmichael) [inline modules/standard/GMP_old.chpl]: I suspect that this file should not be part of the PR.  It can be retrieved from the repo if required.

File: test/modules/standard/gmp/ferguson/gmp_dist_array.chpl [hunk 10]
```diff
@@ -44,16 +41,27 @@ modulus.nextprime(modulus);
 writeln(modulus);
 sum.mod(sum, modulus);
 writeln(sum);
-delete modulus;
-delete sum;
+
 
 // Release the array elements.
+// TODO: remove this section once the records are being released correctly.
```
Comment (noakesmichael) [inline test/modules/standard/gmp/ferguson/gmp_dist_array.chpl]: For face-to-face.

Seems strange that this application needs to think about memory management.

File: modules/standard/GMP_old.chpl (hunk 9 above)
Comment (Ldelaney) [inline modules/standard/GMP_old.chpl]: yup. That would be my bad for not remembering that I had included it in my commit

File: test/modules/standard/gmp/ldelaney/gmp_BigInt_arithmetic.chpl [hunk 11]
```diff
@@ -0,0 +1,40 @@
+use GMP;
+
+// test the arithmetic functions
```
Comment (noakesmichael) [inline test/modules/standard/gmp/ldelaney/gmp_BigInt_arithmetic.chpl]: For face-to-face

I need to think more about the implications of these examples.  Is this really the API we want?
Was there other discussion about this vs. the alternatives that I am unaware of?

File: test/modules/standard/gmp/ferguson/gmp_dist_array.chpl (hunk 10 above)
Comment (Ldelaney) [inline test/modules/standard/gmp/ferguson/gmp_dist_array.chpl]: the distributed array leaks all of its array elements. Since the elements are leaked, the records don't go out of scope and their destructors aren't called. Manually calling the destructors significantly reduces the amount of memory leaked. 

Ideally, it wouldn't need to think about the memory management, and ideally it wouldn't be leaking the array elements. In the meantime, manually freeing the records means that we leak around 2000x less data from this one test.

Comment (noakesmichael) [inline test/modules/standard/gmp/ferguson/gmp_dist_array.chpl]: It’s well known that distributed arrays leak their data and there is an
existing project to tackle this.  We won’t want to add memory management
to other types to work around the current problems issues in distributed arrays.

Mike



On Jul 28, 2016, at 11:59 AM, Laura Delaney <notifications@github.com<mailto:notifications@github.com>> wrote:


In test/modules/standard/gmp/ferguson/gmp_dist_array.chpl<https://github.com/chapel-lang/chapel/pull/4243#discussion_r72681703>:

>
>  // Release the array elements.
> +// TODO: remove this section once the records are being released correctly.


the distributed array leaks all of its array elements. Since the elements are leaked, the records don't go out of scope and their destructors aren't called. Manually calling the destructors significantly reduces the amount of memory leaked.

Ideally, it wouldn't need to think about the memory management, and ideally it wouldn't be leaking the array elements. In the meantime, manually freeing the records means that we leak around 2000x less data from this one test.

Comment (Ldelaney) [inline test/modules/standard/gmp/ferguson/gmp_dist_array.chpl]: I'm not adding manual_free to the BigInt type module, just to that one test so that it leaks less (it still leaks 2000 bytes of other data related to leaking the array elements). 

If we would prefer to let the test leak everything, we could do that.

Comment (noakesmichael) [inline test/modules/standard/gmp/ferguson/gmp_dist_array.chpl]: Oh right.  That was dumb on my part.

Tough call.  I believe I would choose to let this test leak then
but I can understand someone taking the opposing point of
view.


In the original version of the app, the app was responsible for
memory management.  In the new version it’s not supposed
to be.


We have other leaks in other apps right now that are are due to
lower-level bugs.   In every case that I know about we’ve elected
to leave the leak in the App.


I know of one sort-of exception to this rule and was a leak in a
popular library that used inherited a leak.  We chose to modify
the library to cover the leak until we can fix the leak.

Mike







On Jul 28, 2016, at 2:46 PM, Laura Delaney <notifications@github.com<mailto:notifications@github.com>> wrote:


In test/modules/standard/gmp/ferguson/gmp_dist_array.chpl<https://github.com/chapel-lang/chapel/pull/4243#discussion_r72709196>:

>
>  // Release the array elements.
> +// TODO: remove this section once the records are being released correctly.


I'm not adding manual_free to the BigInt type module, just to that one test so that it leaks less (it still leaks 2000 bytes of other data related to leaking the array elements).

If we would prefer to let the test leak everything, we could do that.

Comment (Ldelaney) [inline test/modules/standard/gmp/ferguson/gmp_dist_array.chpl]: I chose to ‘fix’ the leak because:

1)       It still leaks other things, and the information from memleaks correctly guides a developer to the cause of the leak.

a.       I am reasonably sure that when the elements are not leaked, the records will destroy themselves properly.

b.       This will then cause the test to fail (trying to double-free the mpz) which will be loud and noticeable.

2)       The change in scale of what was leaking (on the order of millions of bytes instead of thousands) looked significant.

3)       I could leave a comment explaining that the cleanup is temporary and specifically for this test

File: modules/standard/GMP.chpl (hunk 2 above)
Comment (Ldelaney) [inline modules/standard/GMP.chpl]: This is a documentation comment, and there are six others in the same style which just label the GMP types as their Chapel versions are defined. I'm not sure how I would expand the comment to be more descriptive, and I don't think it makes sense to remove it

File: modules/standard/GMP.chpl (hunk 5 above)
Comment (Ldelaney) [inline modules/standard/GMP.chpl]: yes

File: modules/standard/GMP.chpl (hunk 8 above)
Comment (Ldelaney) [inline modules/standard/GMP.chpl]: In response to 1):
Yes. They should be const ref. I just also have to change the extern declaration to permit that. 

In response to 2): 
How I have the named function interface set up is that for big ints a, b, and c, doing a.someFunction(b, c) should set a to the result of using that function on b and c.
